### ISSUE TITLE
C# API Training Issue

### ISSUE THREAD

**Issue by debadeepta** [Microsoft/CNTK#2560]
Hi,

I am trying to use the recently released C# training API to demonstrate a simple DQN working with Project Malmo, Windows 10, 64-bit, GPU version through Nuget (version 2.2).

Below is the code for an AgentDQN class which defines a simple DQN agent. Everything is working fine except at the point where I am calling the TrainMiniBatch function on the behavior network. It throws an exception: __System.AccessViolationException: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.'__

I have double checked that all the Value items in the <Variable, Value> item have expected data values. I am quite new to CNTK in general and the C# API in particular. Any help will be greatly appreciated. It is quite possible that I am calling something wrong trivially. Happy to send along the entire solution if that will be useful.

`
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Drawing;

using CNTK;

namespace RLMalmo
{
    public class AgentDQN
    {
        private ReplayMemory<Bitmap, int> mReplayMemory;
        private LinearEpsilonAnnealingExplorer mLinearEpsilonAnnealingExplorer;
        private History mHistory;
        private int mNumStepsToUpdateTargetNetwork;

        private int mNumEpisodes;
        private int mReplayMemorySize;
        private int mHistoryLength;
        private double mExplorationStartValue;
        private double mExplorationStopValue;
        private int mExplorationTotalSteps;

        private int mCurrentEpisodeStep;
        private int mCurrentGlobalStep;

        private Function mBehaviorNetwork;
        private Function mTargetNetwork;
        private Function mBehaviorNetworkLoss;
        private Trainer mBehaviorTrainer;

        private DeviceDescriptor mDevice;
        private Variable mTargetQs;
        private Variable mInputs;
        private Variable mActionsOneHot;
        private Variable mPredQs;

        private int mFrameWidth;
        private int mFrameHeight;

        private int mNumActions;
        private int mNumInputChannels;
        private int mNumInputChannelsPerFrame;
        private int mTrainAfterSteps;
        private int mTrainInterval;
        private int mReplayMemoryBatchSize;
        private float mDiscountFactor;

        private Random mRnd;

        public AgentDQN(int numEpisodes, 
                        int replayMemoryMaxSize, 
                        int historyLength, 
                        double explorationStartValue, 
                        double explorationStopValue, 
                        int explorationTotalSteps,
                        int frameWidth,
                        int frameHeight,
                        int numInputChannelsPerFrame,
                        int numActions,
                        int trainAfterSteps,
                        int trainInterval,
                        int replayMemoryBatchSize,
                        float discountFactor,
                        int numStepsToUpdateTargetNetwork)
        {
            mNumEpisodes = numEpisodes;
            mReplayMemorySize = replayMemoryMaxSize;
            mHistoryLength = historyLength;
            mExplorationStartValue = explorationStartValue;
            mExplorationStopValue = explorationStopValue;
            mExplorationTotalSteps = explorationTotalSteps;
            mFrameWidth = frameWidth;
            mFrameHeight = frameHeight;
            mNumInputChannelsPerFrame = numInputChannelsPerFrame;
            mTrainAfterSteps = trainAfterSteps;
            mTrainInterval = trainInterval;
            mReplayMemoryBatchSize = replayMemoryBatchSize;
            mDiscountFactor = discountFactor;
            mNumActions = numActions;
            mNumStepsToUpdateTargetNetwork = numStepsToUpdateTargetNetwork;

            mCurrentEpisodeStep = 0;
            mCurrentGlobalStep = 0;

            mReplayMemory = new ReplayMemory<Bitmap, int>(mReplayMemorySize, mHistoryLength);
            mLinearEpsilonAnnealingExplorer = new LinearEpsilonAnnealingExplorer(mExplorationStartValue, 
                                                                                 mExplorationStopValue, 
                                                                                 mExplorationTotalSteps);

            // Instantiate the input and target variable
            mNumInputChannels = mHistoryLength + 1;
            mInputs = Variable.InputVariable(new int[] { mFrameHeight, mFrameWidth, mNumInputChannels }, DataType.Float);
            mTargetQs = Variable.InputVariable(new int[] { mNumActions }, DataType.Float);
            mActionsOneHot = Variable.InputVariable(new int[] { mNumActions }, DataType.Float);
            mPredQs = Variable.InputVariable(new int[] { mNumActions }, DataType.Float);

            // Define the behavior network
            mDevice = DeviceDescriptor.GPUDevice(0);
            mBehaviorNetwork = BehaviorNetworkDefine();
            mBehaviorNetworkLoss = BehaviorNetworkDefineLoss(mPredQs, mTargetQs, mActionsOneHot);
            mBehaviorTrainer = BehaviorNetworkTrainSetup(mBehaviorNetworkLoss, mBehaviorNetworkLoss);

            // Clone the behavior network to a target network
            mTargetNetwork = mBehaviorNetwork.Clone(ParameterCloningMethod.Freeze);

            // History network
            mHistory = new History(mHistoryLength, mFrameWidth, mFrameHeight, mNumInputChannelsPerFrame);

            // Initialize the random number generator
            mRnd = new Random();
        }

        public int Act(Bitmap state)
        {
            // Bookkeeping
            mCurrentEpisodeStep += 1;
            mCurrentGlobalStep += 1;

            // Should we explore? If so then select random action.
            bool shouldExplore = mLinearEpsilonAnnealingExplorer.ShouldExplore(mCurrentGlobalStep);
            if (shouldExplore)
            {
                int chosenAction = mRnd.Next(mNumActions);
                return chosenAction;
            }

            // Else select argmax over the behavior network's predicted Q-values.
            // -------------------------------------------------------------------

            // Append the state to the short term memory
            mHistory.Append(state);
            // Get state with history
            Value stateWithHistory = mHistory.GetStateWithHistory(mDevice);

            var inputDataMap = new Dictionary<Variable, Value>() { { mInputs, stateWithHistory } };
            var outputDataMap = new Dictionary<Variable, Value>() { { mBehaviorNetwork.Output, null } };
            mBehaviorNetwork.Evaluate(inputDataMap, outputDataMap, mDevice);
            var outputValue = outputDataMap[mBehaviorNetwork.Output];
            IList<IList<float>> actualActionValues = outputValue.GetDenseData<float>(mBehaviorNetwork.Output);
            // For us actualArgMax should be a list of length 1 since 
            // we are giving a single state with history as input.
            var actualArgMax = actualActionValues.Select((IList<float> l) => l.IndexOf(l.Max())).ToList();
            return actualArgMax[0];
        }

        public List<Value> ConvertStateListToValueList(List<List<Bitmap>> states)
        {
            List<Value> statesValue = new List<Value>();
            for (int i = 0; i < states.Count; i++)
            {
                Value thisStateWithHistory = Utils.ConvertSingleStateWithHistoryToValue(states[i], mNumInputChannels, mDevice);
                statesValue.Add(thisStateWithHistory);
            }

            return statesValue;
        }

        public void Train()
        {
            if (mCurrentGlobalStep > mTrainAfterSteps)
            {
                if(mCurrentGlobalStep % mTrainInterval == 0)
                {
                    // Sample random minibatch from replay memory.
                    List<List<Bitmap>> preStates;
                    List<List<Bitmap>> postStates;
                    List<int> actions;
                    List<double> rewards;
                    List<bool> terminals;
                    
                    (preStates, postStates, actions, rewards, terminals) =  mReplayMemory.SampleMinibatchFromReplayMemory(mReplayMemoryBatchSize);

                    // Compute the target values.
                    // TODO This part can be completely vectorized. Will be 
                    // much faster than invoking the network for every single
                    // data point.
                    List<Value> postStatesValue = ConvertStateListToValueList(postStates);

                    // Targets store, initialize to 0
                    float[,] targets = new float[mReplayMemoryBatchSize, mNumActions];
                    for (int i = 0; i < mReplayMemoryBatchSize; i++)
                    {
                        for (int j = 0; j < mNumActions; j++)
                        {
                            targets[i, j] = 0.0f;
                        }
                    }

                    for(int i = 0; i < preStates.Count; i++)
                    {
                        if (terminals[i])
                        {
                            // TODO Make sure that when a state is 
                            // terminal its corresponding reward is also correct
                            targets[i, actions[i]] = (float)rewards[i];
                        }
                        else
                        {
                            // Get the max action values for the postStates 
                            // using the target network
                            var inputDataMap = new Dictionary<Variable, Value>() { { mTargetNetwork.Arguments[0], postStatesValue[i] } };
                            var outputDataMap = new Dictionary<Variable, Value>() { { mTargetNetwork.Output, null } };
                            mTargetNetwork.Evaluate(inputDataMap, outputDataMap, mDevice);
                            var outputValue = outputDataMap[mTargetNetwork.Output];
                            IList<IList<float>> actualActionValues = outputValue.GetDenseData<float>(mTargetNetwork.Output);
                            // For us actualMax should be a list of length 1 since 
                            // we are giving a single state with history as input.
                            var actualMax = actualActionValues.Select((IList<float> l) => l.Max());
                            targets[i, actions[i]] = (float)rewards[i] + mDiscountFactor * actualMax.ToList()[0];
                        }
                    }

                    Console.WriteLine("Here 1");

                    // Flatten target array
                    float[] targetsFlat = new float[mReplayMemoryBatchSize * mNumActions];
                    int counter = 0;
                    for (int i = 0; i < mReplayMemoryBatchSize; i++)
                    {
                        for (int j = 0; j < mNumActions; j++)
                        {
                            targetsFlat[counter] = targets[i, j];
                            counter += 1;
                        }
                    }

                    // Compute the predicted Q-values for the preStates
                    Value preStatesValueBatch = Utils.ConvertListOfStatesWithHistoryToValue(preStates, mNumInputChannels, mDevice);
                    var inputDataMapPreds = new Dictionary<Variable, Value>() { { mInputs, preStatesValueBatch} };
                    var outputDataMapPreds = new Dictionary<Variable, Value>() { { mBehaviorNetwork.Output, null } };
                    mBehaviorNetwork.Evaluate(inputDataMapPreds, outputDataMapPreds, mDevice);
                    var outputQPredsValue = outputDataMapPreds[mBehaviorNetwork.Output];
                    var outputQPredsActualValues = outputQPredsValue.GetDenseData<float>(mBehaviorNetwork.Output);

                    Console.WriteLine("Here 2");

                    // Update the behavior policy's parameters by doing a step of
                    // gradient descent.
                    //--------------------------------------------------------------

                    // First get the actions in one-hot Value format
                    Value actionsOneHotValueBatch = Value.CreateBatch<float>(mNumActions, actions, mDevice);
                    
                    // Get the targets in Value format
                    Value targetsValueBatch = Value.CreateBatch<float>(new int[] { mNumActions }, targetsFlat, mDevice);

                    // Train
                    Dictionary<Variable, Value> argDict = new Dictionary<Variable, Value>
                    {
                        { mInputs, preStatesValueBatch },
                        { mActionsOneHot, actionsOneHotValueBatch },
                        { mTargetQs, targetsValueBatch },
                        { mPredQs, outputQPredsValue }
                    };
                    mBehaviorTrainer.TrainMinibatch(argDict, mDevice);
                    Console.WriteLine("Trained a minibatch.");
                }

                // If step mod mNumStepsToUpdateTargetNetwork == 0, then update
                // the target network to be the same as the behavior network. 
                if (mCurrentEpisodeStep % mNumStepsToUpdateTargetNetwork == 0)
                {
                    mTargetNetwork = mBehaviorNetwork.Clone(ParameterCloningMethod.Freeze);
                }

            }
        }

        public void Observe(Bitmap oldState, int action, double reward, bool done)
        {
            if (done)
            {
                // Reset the short term memory
                mHistory.Reset();
            }

            // Append to replay buffer
            mReplayMemory.Append(oldState, action, reward, done);
        }

        public Function BehaviorNetworkDefine()
        {
            // The first few convolution layers
            float convWScale = 1.0f;
            int kernelWidth1 = 8;
            int kernelHeight1 = 8;

            int kernelWidth2 = 4;
            int kernelHeight2 = 4;

            int kernelWidth3 = 3;
            int kernelHeight3 = 3;

            int outFeatureMapCount1 = 16;
            int outFeatureMapCount2 = 32;
            int outFeatureMapCount3 = 32;

            var convParams1 = new Parameter(new int[] {kernelWidth1, kernelHeight1, mNumInputChannels, outFeatureMapCount1}, DataType.Float, CNTKLib.GlorotUniformInitializer(convWScale, -1, 2), mDevice);
            var convFunction1 = CNTKLib.ReLU(CNTKLib.Convolution(convParams1, mInputs, new int[] {4, 4, mNumInputChannels}));

            var convParams2 = new Parameter(new int[] {kernelWidth2, kernelHeight2, outFeatureMapCount1, outFeatureMapCount2}, DataType.Float, CNTKLib.GlorotUniformInitializer(convWScale, -1, 2), mDevice);
            var convFunction2 = CNTKLib.ReLU(CNTKLib.Convolution(convParams2, convFunction1, new int[] {2, 2, outFeatureMapCount1}));

            var convParams3 = new Parameter(new int[] {kernelWidth3, kernelHeight3, outFeatureMapCount2, outFeatureMapCount3}, DataType.Float, CNTKLib.GlorotUniformInitializer(convWScale, -1, 2), mDevice);
            var convFunction3 = CNTKLib.ReLU(CNTKLib.Convolution(convParams3, convFunction2, new int[] {1, 1, outFeatureMapCount2}));

            // Dense layers
            int numDenseUnits = 256;
            Activation activation = Activation.ReLU;
            var nonlinear1 = Utils.Dense(convFunction3, numDenseUnits, mDevice, activation, "firstDenseLayer");
            var nonlinear2 = Utils.Dense(nonlinear1, mNumActions, mDevice, activation, "firstDenseLayer");

            return nonlinear2;
        }

        public Function BehaviorNetworkDefineLoss(Variable prediction, Variable targets, Variable actionTakenOneHot)
        {
            // Have to suppress the contribution of actions which were
            // not taken at that point. One way is to multiply the prediction 
            // with the action mask in one-hot format and then reduce_sum 
            // as done in the criterion function here:
            // https://github.com/Microsoft/CNTK/blob/master/Examples/ReinforcementLearning/DeepQNeuralNetwork.py
            var maskedPrediction = CNTKLib.ElementTimes(prediction, actionTakenOneHot);
            // TODO Do I need a reduce_sum op or will SquaredError handle that?
            // Will find out during compilation or runtime.
            var loss = CNTKLib.SquaredError(maskedPrediction, targets);
            return loss;
        }

        public Trainer BehaviorNetworkTrainSetup(Function loss, Function evalError)
        {
            CNTK.TrainingParameterScheduleDouble learningRatePerSample = new CNTK.TrainingParameterScheduleDouble(0.02);
            IList<Learner> parameterLearners = new List<Learner>() { Learner.SGDLearner(mBehaviorNetwork.Parameters(), learningRatePerSample) };
            var trainer = Trainer.CreateTrainer(mBehaviorNetwork, loss, evalError, parameterLearners);
            return trainer;
        }

    }
}

`

**Comment by debadeepta** [comment]
I figured out what was going on wrong. I was creating the prediction values myself and feeding them into the train minibatch call via a dictionary. That is unnecessary since train minibatch call will do that itself.
